plugins/genesys-gl32xx/fu-genesys-gl32xx-device.c

```diff
@@ -0,0 +1,879 @@
+	g_autoptr(GByteArray) buf = NULL;
+
+	buf = fu_genesys_gl32xx_device_cmd_get_version(self, error);
+	if (buf == NULL) {
+		g_prefix_error(error, "failed to read version: ");
+		return FALSE;
+	}
+	if (buf->len < 0x24) {
+		g_set_error_literal(error,
+				    G_IO_ERROR,
+				    G_IO_ERROR_NOT_SUPPORTED,
+				    "failed to read version");
+		return FALSE;
+	}
+
+	version = g_strdup_printf("%c%c%c%c",
+				  buf->data[0x20],
+				  buf->data[0x21],
+				  buf->data[0x22],
+				  buf->data[0x23]);
+	fu_device_set_version(FU_DEVICE(self), version);
+
+	return TRUE;
+}
+
+static gboolean
+fu_genesys_gl32xx_device_check_rom_mode(FuGenesysGl32xxDevice *self,
+					guint8 *cmd,
+					gsize cmdsz,
+					GError **error)
+{
+	const guint8 ext_rom_mode[] = {0x58, 0x52, 0x4F, 0x4D}; /* "XROM" */
+	const guint8 int_rom_mode[] = {0x49, 0x4E, 0x54, 0x2D}; /* "INT-" */
+	guint8 data[4] = {0};
+
+	if (!fu_genesys_gl32xx_device_send_scsi_cmd_v3(self,
+						       cmd,
+						       cmdsz,
+						       data,
+						       sizeof(data),
+						       FU_GENESYS_GL32XX_DIRECTION_IN,
+						       error))
+		return FALSE;
+	if (fu_memcmp_safe(int_rom_mode, sizeof(int_rom_mode), data, sizeof(data), NULL)) {
+		fu_device_add_flag(FU_DEVICE(self), FWUPD_DEVICE_FLAG_IS_BOOTLOADER);
+		return TRUE;
+	}
+	if (fu_memcmp_safe(ext_rom_mode, sizeof(ext_rom_mode), data, sizeof(data), NULL)) {
+		fu_device_remove_flag(FU_DEVICE(self), FWUPD_DEVICE_FLAG_IS_BOOTLOADER);
+		return TRUE;
+	}
+
+	g_set_error_literal(error,
+			    FWUPD_ERROR,
+			    FWUPD_ERROR_NOT_FOUND,
+			    "no supported devices detected");
+	return FALSE;
+}
+
+/* safe to call in any mode */
+static gboolean
+fu_genesys_gl32xx_device_ensure_rom_mode(FuGenesysGl32xxDevice *self, GError **error)
+{
+	guint8 cmd_gl323x[] = {0xE4, 0x01, 0x00, 0xDC, 0x04, 0x00};
+	guint8 cmd_gl3224[] = {0xE4, 0x01, 0x00, 0xFC, 0x04, 0x00};
+	g_autoptr(GError) error_local = NULL;
+
+	/* check for 3230, 3231, 3232, 3230S, 3231S, 3232S first */
+	/* ignore error here */
+	if (!fu_genesys_gl32xx_device_check_rom_mode(self,
+						     cmd_gl323x,
+						     sizeof(cmd_gl323x),
+						     &error_local)) {
+		g_debug("ignoring: %s", error_local->message);
+	} else {
+		fu_genesys_gl32xx_device_set_chip_name(self, "GL323x");
+		return TRUE;
+	}
+
+	/* check the 3224 */
+	if (!fu_genesys_gl32xx_device_check_rom_mode(self, cmd_gl3224, sizeof(cmd_gl3224), error))
+		return FALSE;
+
+	fu_genesys_gl32xx_device_set_chip_name(self, "GL3224");
+	return TRUE;
+}
+
+static gboolean
+fu_genesys_gl32xx_device_verify_chip_id(FuGenesysGl32xxDevice *self, GError **error)
+{
+	struct {
+		const gchar *name;
+		guint8 id[3];
+		guint32 idsz;
+		guint8 cmd_req[6];
+		guint8 data_req[4];
+		guint32 data_reqsz;
+		guint8 cmd_get[6];
+	} const flash_type[3] = {
+	    {
+		.name = "gd25d10b",
+		.id = {0xC8, 0x10},
+		.idsz = 2,
+		.cmd_req = {0xF3, 0x02, 0x00, 0x04, 0x00, 0x02},
+		.data_req = {0x90, 0x00, 0x00, 0x00},
+		.data_reqsz = 4,
```

Comment:
if you change this to 0x90 and `.data_reqsz = 1,` does it still work? I've never seen a 4 byte ID request before.

Reply:
from publicly available manual of `gd25d10b` chip:
_The command is initiated by driving the CS# pin low and shifting the command code “90H” followed by a 24-bit
address (A23-A0) of 000000H._

however it's a bit strange why `90H` is used for `gd25d10b` while the doc shows it is supporting `9F` command as well :thinking:

Reply:
Sometimes 90H returns different data to 9F-- but everything produced in the last 5 years seems to support 9F. If 9F works I'd really like to keep things simple.